o1vm/src/mips/constraints.rs

```diff
@@ -482,67 +493,66 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
-            // in this row
+            // +4 constraints
+            // Constrain the bytes flags depending on the number of preimage
+            // bytes read in this row
             {
                 // When at least has_1_byte, then any number of bytes can be
                 // read <=> Check that you can only read 1, 2, 3 or 4 bytes
                 self.add_constraint(
                     has_n_bytes[0].clone()
-                        * (num_read_bytes.clone() - Expr::from(1))
-                        * (num_read_bytes.clone() - Expr::from(2))
-                        * (num_read_bytes.clone() - Expr::from(3))
-                        * (num_read_bytes.clone() - Expr::from(4)),
+                        * (num_preimage_bytes_read.clone() - Expr::from(1))
+                        * (num_preimage_bytes_read.clone() - Expr::from(2))
+                        * (num_preimage_bytes_read.clone() - Expr::from(3))
+                        * (num_preimage_bytes_read.clone() - Expr::from(4)),
                 );
 
                 // When at least has_2_byte, then any number of bytes can be
-                // read except 1
-                self.constraints.push(
+                // read from the preimage except 1
+                self.add_constraint(
                     has_n_bytes[1].clone()
-                        * (num_read_bytes.clone() - Expr::from(2))
-                        * (num_read_bytes.clone() - Expr::from(3))
-                        * (num_read_bytes.clone() - Expr::from(4)),
+                        * (num_preimage_bytes_read.clone() - Expr::from(2))
+                        * (num_preimage_bytes_read.clone() - Expr::from(3))
+                        * (num_preimage_bytes_read.clone() - Expr::from(4)),
                 );
                 // When at least has_3_byte, then any number of bytes can be
-                // read except 1 nor 2
-                self.constraints.push(
+                // read from the preimage except 1 nor 2
+                self.add_constraint(
                     has_n_bytes[2].clone()
-                        * (num_read_bytes.clone() - Expr::from(3))
-                        * (num_read_bytes.clone() - Expr::from(4)),
+                        * (num_preimage_bytes_read.clone() - Expr::from(3))
+                        * (num_preimage_bytes_read.clone() - Expr::from(4)),
                 );
 
-                // When has_4_byte, then only can read 4
-                self.constraints
-                    .push(has_n_bytes[3].clone() * (num_read_bytes.clone() - Expr::from(4)));
+                // When has_4_byte, then only can read 4 preimage bytes
+                self.add_constraint(
+                    has_n_bytes[3].clone() * (num_preimage_bytes_read.clone() - Expr::from(4)),
+                );
             }
         }
 
+        // FIXED LOOKUPS
+        // Check that 0 <= preimage read <= actual read <= len <= 4
+        self.add_lookup(Lookup::write_one(
```

Comment:
Note (reminder for me): good that we check. We must check that `len`, a value coming from a register, is smaller than 4 because we do not do it in any other place.